### PR TITLE
Rebase old specs branch, drop some old APIs, describe allowed test.config options

### PR DESCRIPTION
**Pull request by lavrin**
esl/mongooseim#477 checks backward compatibility and any necessary test suite fixes.

### CODE REVIEW

File: src/escalus_bosh.erl [hunk 1]
```diff
@@ -497,7 +497,7 @@ detect_type(Attrs) ->
         {_,         Version} -> {streamstart,Version}
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space after "," on line 497

File: src/escalus_users.erl [hunk 2]
```diff
@@ -203,34 +219,24 @@ update_userspec(Config, UserName, Option, Value) ->
...
-    Users;
-get_users({by_name, Names}, Users) ->
-    [get_user_by_name(Name, Users) || Name <- Names].
-
--spec get_user_by_name(atom()) -> spec().
-get_user_by_name(Name) ->
-    get_user_by_name(Name, get_users(all)).
-
--spec get_user_by_name(atom(), escalus:config()) -> {atom(), escalus:config()}.
+-spec get_user_by_name(user_name(), escalus:config()) -> {user_name(), escalus:config()}.
```
Comment (gadgetci) [inline src/escalus_users.erl]: According to **Elvis**:
> Line 229 is too long: -spec get_user_by_name(user_name(), escalus:config()) -> {user_name(), escalus:config()}..

File: src/escalus_users.erl [hunk 3]
```diff
@@ -253,6 +260,8 @@ verify_creation({error, Error, Raw}) ->
     error_logger:error_msg("error when trying to register user: ~s~n", [RawStr]),
```
Comment (gadgetci) [inline src/escalus_users.erl]: According to **Elvis**:
> Line 260 is too long:     error_logger:error_msg("error when trying to register user: ~s~n", [RawStr]),.

File: src/escalus_users.erl [hunk 4]
```diff
@@ -90,27 +89,28 @@ stop(Config) ->
...
-        config -> escalus_config:get_config(escalus_users, Config, []);
-        _      -> get_users(Who)
-    end,
     case auth_type(Config) of
         {escalus_user_db, xmpp} ->
+            Users = get_users(Who),
             [delete_user(Config, User) || User <- Users];
         {escalus_user_db, {module, M, _}} ->
-            M:delete_users(Config, Users)
+            M:delete_users(Config, Who)
```
Comment (gadgetci) [inline src/escalus_users.erl]: According to **Elvis**:
> Remove the dynamic function call on line 113. Only modules that define callbacks should make dynamic calls.

File: src/escalus_client.erl [hunk 5]
```diff
@@ -44,6 +44,8 @@
 %% Public API
 %%--------------------------------------------------------------------
 
+-spec start(escalus:config(), escalus_users:user_spec(), binary()) -> {ok, _}
+                                                                    | {error, _}.
```
Comment (gadgetci) [inline src/escalus_client.erl]: According to **Elvis**:
> Line 48 is too long:                                                                     | {error, _}..

File: src/escalus_bosh.erl [hunk 6]
```diff
@@ -497,7 +497,7 @@ detect_type(Attrs) ->
...
     end.
 
-host_to_list({_,_,_,_} = IP4) -> inet:ntoa(IP4);
-host_to_list({_,_,_,_,_,_,_,_} = IP6) -> inet:ntoa(IP6);
+host_to_list({_,_,_,_} = IP4) -> inet_parse:ntoa(IP4);
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space after "," on line 500

File: src/escalus_users.erl [hunk 7]
```diff
@@ -90,27 +89,28 @@ stop(Config) ->
...
     case auth_type(Config) of
         {escalus_user_db, xmpp} ->
-            CreationResults = [create_user(Config, User) || User <- Users],
-            lists:foreach(fun verify_creation/1, CreationResults);
+            create_users_via_xmpp(Config, Who);
         {escalus_user_db, {module, M, _}} ->
-            M:create_users(Config, Users)
-    end,
-    [{escalus_users, Users}] ++ Config.
+            M:create_users(Config, Who)
```
Comment (gadgetci) [inline src/escalus_users.erl]: According to **Elvis**:
> Remove the dynamic function call on line 96. Only modules that define callbacks should make dynamic calls.

File: src/escalus_bosh.erl [hunk 8]
```diff
@@ -497,7 +497,7 @@ detect_type(Attrs) ->
...
+host_to_list({_,_,_,_,_,_,_,_} = IP6) -> inet_parse:ntoa(IP6);
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space after "," on line 501

File: src/escalus_connection.erl [hunk 9]
```diff
@@ -59,8 +59,8 @@
 %%% Public API
 %%%===================================================================
 
--spec start(escalus_users:spec()) -> {ok, client(), escalus_users:spec()} |
-                                     {error, any()}.
+-spec start(escalus_users:user_spec()) -> {ok, client(), escalus_users:user_spec()}
```
Comment (gadgetci) [inline src/escalus_connection.erl]: According to **Elvis**:
> Line 62 is too long: -spec start(escalus_users:user_spec()) -> {ok, client(), escalus_users:user_spec()}.

File: src/escalus_client.erl [hunk 10]
```diff
@@ -44,6 +44,8 @@
...
 start(Config, UserSpec, Resource) ->
     EventClient = escalus_event:new_client(Config, UserSpec, Resource),
     Options = escalus_users:get_options(Config, UserSpec, Resource, EventClient),
```
Comment (gadgetci) [inline src/escalus_client.erl]: According to **Elvis**:
> Line 51 is too long:     Options = escalus_users:get_options(Config, UserSpec, Resource, EventClient),.

File: src/escalus_users.erl [hunk 11]
```diff
@@ -261,7 +270,7 @@ delete_user(Config, {_Name, UserSpec}) ->
     escalus_connection:stop(Conn),
     Result.
 
--spec auth_type(ct:config()) -> {module, atom(), list()} | xmpp.
+-spec auth_type(ct:config()) -> {escalus_user_db, {module, atom(), list()} | xmpp}.
```
Comment (gadgetci) [inline src/escalus_users.erl]: According to **Elvis**:
> Line 273 is too long: -spec auth_type(ct:config()) -> {escalus_user_db, {module, atom(), list()} | xmpp}..

**Comment by michalwski** [comment]
:+1:

**Pull request merged**